File: crates/app/peritus-tui/src/model/chat/catalog.rs
//! Typed local command metadata. Only directly entered commands are executable intents.

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
enum Arguments {
    None,
    Message,
    Model,
    Effort,
    Sessions,
    Fork,
    Queue,
    Context,
    Compact,
    Brief,
    Preview,
    Checkpoint,
    Rewind,
    Attachment,
    Goal,
    Pause,
    Resume,
    Budget,
    Permissions,
    Init,
    Memory,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub(super) enum Command {
    Chat,
    Plan,
    Review,
    Build,
    Model,
    Effort,
    New,
    Status,
    Diff,
    Runs,
    Trace,
    Terminal,
    Approvals,
    Details,
    Stop,
    Accept,
    Commit,
    Export,
    Discard,
    Run,
    Reconnect,
    Doctor,
    Sessions,
    Fork,
    Queue,
    Context,
    Compact,
    Brief,
    Goal,
    Pause,
    Resume,
    Usage,
    Budget,
    Preview,
    Checkpoint,
    Rewind,
    Attach,
    Files,
    Permissions,
    Init,
    Memory,
    Help,
    Quit,
}

mod entries;
use entries::COMMANDS;

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub(super) struct CommandSpec {
    pub(super) command: Command,
    pub(super) name: &'static str,
    pub(super) description: &'static str,
    arguments: Arguments,
}

pub(super) fn parse(text: &str) -> Result<(Command, &str), &'static str> {
    let (name, arguments) = text
        .split_once(char::is_whitespace)
        .map_or((text, ""), |(name, arguments)| (name, arguments.trim()));
    let spec = COMMANDS
        .iter()
        .find(|spec| spec.name == name)
        .ok_or("Unknown slash command. Type / to see available commands.")?;
    if spec.arguments == Arguments::None && !arguments.is_empty() {
        return Err("This command takes no arguments; draft retained.");
    }
    Ok((spec.command, arguments))
}

/// Complete command names without invoking a provider; exact matches come first.
pub(super) fn completions(text: &str) -> Vec<(String, &'static str)> {
    if !text.starts_with('/') {
        return Vec::new();
    }
    if text.contains(char::is_whitespace) {
        return argument_completions(text);
    }
    let mut matches: Vec<_> = COMMANDS
        .iter()
        .filter(|spec| spec.name.starts_with(text))
        .map(|spec| (spec.name.to_owned(), spec.description))
        .collect();
    matches.sort_by_key(|(name, _)| name != text);
    matches
}

fn argument_completions(text: &str) -> Vec<(String, &'static str)> {
    let Some((prefix, partial)) = text.rsplit_once(char::is_whitespace) else { return Vec::new() };
    let words: Vec<_> = prefix.split_whitespace().collect();
    let choices: &[&str] = match words.as_slice() {
        ["/brief"] => &["objective", "acceptance", "constraints", "assumptions"],
        ["/goal"] => &["confirm", "clear", "criterion"],
        ["/goal", "criterion"] => &["graphical", "remove-graphical"],
        ["/goal", "clear"] => &["confirm"],
        ["/pause"] => &["now", "after-operation", "before-edit"],
        ["/budget"] => &["none", "time=", "requests=", "tools=", "tokens="],
        ["/preview"] => &["results", "launch", "play", "capture", "stop", "check", "feedback"],
        ["/context"] => &["next", "history", "show", "more", "previous"],
        ["/queue"] => &[
            "add", "edit", "correct", "hold", "release", "withdraw", "order", "history", "pending",
            "show", "next", "previous", "retry",
        ],
        ["/sessions"] => {
            &["new", "open", "rename", "pin", "unpin", "archive", "unarchive", "retry"]
        }
        ["/fork"] => &["read-only", "isolated"],
        ["/checkpoint"] => &["show"],
        ["/permissions"] => &["restrict", "grant"],
        ["/permissions", "restrict" | "grant"] => &["read", "write", "process", "network"],
        ["/init"] => &["apply", "decline"],
        ["/memory"] => &[
            "save",
            "revise",
            "pin",
            "unpin",
            "project",
            "conversation",
            "forget",
            "history",
            "more",
            "previous",
        ],
        ["/effort"] => &[
            "writer", "reviewer", "fixer", "default", "minimal", "low", "medium", "high", "xhigh",
            "max", "ultra",
        ],
        ["/effort", "writer" | "reviewer" | "fixer"] => {
            &["default", "minimal", "low", "medium", "high", "xhigh", "max", "ultra"]
        }
        ["/model"] => &["writer", "reviewer", "fixer", "refresh", "manual"],
        ["/model", "writer" | "reviewer" | "fixer"] => &["refresh", "manual"],
        _ => &[],
    };
    let mut matches: Vec<_> = choices
        .iter()
        .filter(|choice| choice.starts_with(partial))
        .map(|choice| (format!("{prefix} {choice}"), "Tab completes · Enter submits"))
        .collect();
    matches.sort_by_key(|(candidate, _)| candidate != text);
    matches
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn catalog_is_unique_and_every_entry_dispatches_and_completes() {
        for (index, spec) in COMMANDS.iter().enumerate() {
            assert!(!COMMANDS[..index].iter().any(|prior| prior.name == spec.name));
            assert_eq!(parse(spec.name), Ok((spec.command, "")));
            assert_eq!(
                completions(spec.name).first(),
                Some(&(spec.name.to_owned(), spec.description))
            );
            if spec.arguments == Arguments::None {
                assert!(parse(&format!("{} unexpected", spec.name)).is_err());
            }
        }
        assert!(parse("/unknown").is_err());
        assert!(parse("/STOP").is_err());
        assert!(parse(" /stop").is_err());
        assert!(completions("a /stop").is_empty());
    }

    #[test]
    fn completion_is_argument_aware_and_never_rewrites_freeform_content() {
        for (prefix, expected) in [
            ("/effort reviewer x", "/effort reviewer xhigh"),
            ("/sessions ren", "/sessions rename"),
            ("/model writer man", "/model writer manual"),
        ] {
            assert_eq!(completions(prefix).first().expect("completion").0, expected);
        }
        for text in [
            "/sessions new exact title",
            "/model writer manual arbitrary-id",
            "/chat keep this message",
            "/doctor repair",
        ] {
            assert!(completions(text).is_empty(), "{text}");
        }
    }
}

File: webui/src/lib/commands/catalog.ts
export interface Command { id: string; label: string; detail: string; group: string; slash: string; args?: string }
export const commands: Command[] = [
  {id:'improvements',label:'Improvement inbox',detail:'Collect suggestions; explicitly generate and test selected harness patches',group:'Harness',slash:'/improvements'},
  {id:'new',label:'New conversation',detail:'Open a sibling session in this project',group:'Workspace',slash:'/new'},
  {id:'nest',label:'Nest a conversation',detail:'Add a child beneath the active session',group:'Workspace',slash:'/nest'},
  {id:'open',label:'Open project',detail:'Add another project to this browser tab',group:'Workspace',slash:'/open',args:'Absolute project directory'},
  {id:'files',label:'File explorer',detail:'Browse project files, including hidden and ignored files',group:'Workspace',slash:'/files'},
  {id:'git',label:'Source control',detail:'Branches, remotes, staging, commit, fetch, pull, and push',group:'Workspace',slash:'/git'},
  {id:'settings',label:'Configure console',detail:'Theme, behavior, layout, shortcuts, and dotfile',group:'Workspace',slash:'/settings'},
  {id:'sessions',label:'Session library',detail:'Reopen and organize saved sessions',group:'Workspace',slash:'/sessions'},
  {id:'chat',label:'Chat mode',detail:'Discuss or direct the harness',group:'Conversation',slash:'/chat'},
  {id:'plan',label:'Plan mode',detail:'Read-only planning and discussion',group:'Conversation',slash:'/plan'},
  {id:'review',label:'Review mode',detail:'Independent read-only review',group:'Conversation',slash:'/review'},
  {id:'build',label:'Build mode',detail:'Writer, checks, reviewer, and fixer pipeline',group:'Conversation',slash:'/build'},
  {id:'stop',label:'Stop current work',detail:'Keep effects already completed',group:'Conversation',slash:'/stop'},
  {id:'model',label:'Models and reasoning',detail:'Select models and effort for each role',group:'Conversation',slash:'/model'},
  {id:'effort',label:'Reasoning effort',detail:'Adjust per-role reasoning depth',group:'Conversation',slash:'/effort'},
  {id:'details',label:'Activity details',detail:'Inspect public tool and status observations',group:'Conversation',slash:'/details'},
  {id:'status',label:'Current run status',detail:'Inspect the observed daemon state',group:'Conversation',slash:'/status'},
  {id:'diff',label:'Candidate diff',detail:'Review exact candidate changes',group:'Results',slash:'/diff'},
  {id:'runs',label:'Run history',detail:'Inspect and resume daemon-owned runs',group:'Results',slash:'/runs'},
  {id:'retry',label:'Retry current run',detail:'Resume a failed or stopped run',group:'Results',slash:'/retry'},
  {id:'export',label:'Export candidate',detail:'Export the exact candidate patch',group:'Results',slash:'/export'},
  {id:'accept',label:'Accept candidate',detail:'Open the exact-run CLI acceptance flow',group:'Results',slash:'/accept'},
  {id:'commit',label:'Commit candidate',detail:'Commit the managed candidate; distinct from Git commit',group:'Results',slash:'/commit'},
  {id:'discard',label:'Discard candidate',detail:'Remove this run’s candidate changes',group:'Results',slash:'/discard'},
  {id:'consoles',label:'Retained consoles',detail:'Reopen or terminate this project’s CLI processes',group:'Harness',slash:'/consoles'},
  {id:'terminal',label:'Full harness console',detail:'Interactive Peritus CLI with keyboard and mouse controls',group:'Harness',slash:'/terminal'},
  {id:'providers',label:'Provider setup',detail:'Account login, API routes, models, and capability tests',group:'Harness',slash:'/providers'},
  {id:'workspaces',label:'Workspace setup',detail:'Register, trust, repair, or forget a workspace',group:'Harness',slash:'/workspaces'},
  {id:'reconnect',label:'Reconnect',detail:'Reload workspace and inspect original operation receipts',group:'Harness',slash:'/reconnect'},
  {id:'update',label:'Update Peritus',detail:'Run the interactive product update',group:'Harness',slash:'/update'},
  ...[
    ['approvals','Approvals','Inspect and answer pending approvals',''],['doctor','Diagnostics','Inspect local prerequisites',''],
    ['trace','Execution trace','Inspect the current execution trace',''],['queue','Input queue','Add, hold, release, edit, or withdraw queued input','add <message>'],
    ['context','Model context','Inspect eligible input and sealed context manifests','next'],['compact','Compact context','Preview and confirm a local compaction',''],
    ['brief','Project brief','Inspect or update objective, acceptance, and constraints','objective <text>'],['goal','Persistent goal','Define criteria and a bounded goal','<objective>'],
    ['pause','Pause goal','Pause at a selected safe boundary','after-operation'],['resume','Resume goal','Resume a paused persistent goal',''],
    ['usage','Usage accounting','Inspect measured requests, tools, time, and tokens',''],['budget','Execution budget','Set time, request, tool, and token limits','time=30m'],
    ['preview','Result preview','Launch, capture, check, and annotate results','results'],['checkpoint','Checkpoint','Create or inspect a durable checkpoint','<name>'],
    ['rewind','Rewind','Preview exact restore effects before confirmation','<checkpoint>'],['attach','Attach context','Explicitly attach a file or supported image','<path>'],
    ['permissions','Workspace permissions','Inspect and restrict or grant capabilities',''],['init','Project initialization','Discover local controls and preview instruction files',''],
    ['memory','Working memory','Save, revise, scope, pin, and forget guidance',''],['fork','Fork conversation','Create a read-only or isolated branch','read-only'],
    ['run','Execute candidate','Run the candidate’s recorded instructions',''],
  ].map(([id,label,detail,args]) => ({ id:id!,label:label!,detail:detail!,args:args!,group:'Workbench console',slash:`/${id}` })),
  {id:'cli',label:'Scriptable CLI command',detail:'Artifacts, prompts, events, terminal attachments, shutdown, and completions',group:'Harness',slash:'/cli',args:'status'},
  {id:'help',label:'Command directory',detail:'Every function, with a mouse action and slash command',group:'Workspace',slash:'/help'},
  {id:'close-project',label:'Close project',detail:'Close this project tab and retain its sessions',group:'Workspace',slash:'/close-project'},
];
export function matchesShortcut(event: KeyboardEvent, binding: string): boolean {
  const parts = binding.toLowerCase().split('+'); const key = parts.pop();
  return event.key.toLowerCase() === key && (event.metaKey || event.ctrlKey) === parts.includes('mod') && event.shiftKey === parts.includes('shift') && event.altKey === parts.includes('alt');
}
export function formatShortcut(binding: string | undefined, apple: boolean): string {
  const names: Record<string,string> = {mod:apple?'Cmd':'Ctrl',alt:apple?'Option':'Alt',shift:'Shift'};
  return binding?.split('+').map(part=>names[part.toLowerCase()]??(part.length===1?part.toUpperCase():part)).join(' ')??'';
}
